core/optaplanner-core-impl/src/test/java/org/optaplanner/core/impl/partitionedsearch/DefaultPartitionedSearchPhaseTest.java

```diff
@@ -184,4 +185,54 @@ void shutdownMainThreadAbruptly() throws InterruptedException {
                 .hasRootCauseExactlyInstanceOf(InterruptedException.class);
     }
 
+    @Test
+    @Timeout(60)
+    // All the credits to https://github.com/BobbyHirst.
+    void solvePartitionedWithProblemChange() throws InterruptedException {
+        // Create a partitioned daemon solver.
+        SolverFactory<TestdataSolution> solverFactory = createSolverFactory(true, SolverConfig.MOVE_THREAD_COUNT_NONE,
+                1);
+        Solver<TestdataSolution> solver = solverFactory.buildSolver();
+
+        final int valueCount = 4;
+        TestdataSolution solution = TestdataSolution.generateSolution(valueCount, valueCount);
+
+        AtomicReference<TestdataSolution> bestSolution = new AtomicReference<>();
+        CountDownLatch solutionWithProblemChangeReceived = new CountDownLatch(1);
+
+        CountDownLatch solvingStarted = new CountDownLatch(1);
+        ((DefaultSolver<TestdataSolution>) solver).addPhaseLifecycleListener(
+                new PhaseLifecycleListenerAdapter<>() {
+                    @Override
+                    public void solvingStarted(SolverScope<TestdataSolution> solverScope) {
+                        solvingStarted.countDown();
+                    }
+                });
+
+        solver.addEventListener(bestSolutionChangedEvent -> {
+            if (bestSolutionChangedEvent.isEveryProblemChangeProcessed()) {
+                TestdataSolution newBestSolution = bestSolutionChangedEvent.getNewBestSolution();
+                if (newBestSolution.getValueList().size() == valueCount + 1) {
+                    bestSolution.set(newBestSolution);
+                    solutionWithProblemChangeReceived.countDown();
+                }
+            }
+        });
+
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+        executorService.submit(() -> {
+            solver.solve(solution);
+        });
+
+        solvingStarted.await(); // Make sure we submit a ProblemChange only after the Solver started solving.
+        solver.addProblemChange((workingSolution, problemChangeDirector) -> {
+            problemChangeDirector.addProblemFact(new TestdataValue("added value"), solution.getValueList()::add);
+        });
+
+        solutionWithProblemChangeReceived.await();
+        assertThat(bestSolution.get().getValueList()).hasSize(valueCount + 1);
```

Comment:
Failure here will not shut down the executor.

Reply:
Good catch, thanks